### DISCUSSION TITLE
*open source* anti cheat using zero knowledge proofs cryptography

### DISCUSSION THREAD

**Discussion by gogo2464** [godotengine/godot-proposals#7619] [General Discussions]
Each modern anti cheat still use code obfuscation in order to bypass patching and packet forgery in online games. 

The issue is: for open source philosophy video games such as homebrew, it is absolutely impossibe and they have to choose between a dilemna between:

1. accept cheaters
2. abandon the freedom of free software in order to ... *limit* ... the cheaters numbers. There will still remain cheaters anyway.

We might protect the game against cheaters hidding data remainning encrypted.

I suggest to develop a mecanism of protection against video game cheating using zero proof knowledge in godot in order to entierely block:

- stathack
- wallhack

I do not know if this could work against aimbot.

It will never ever works against farm bots. We may need a boring captcha against this threat.

I do not know how much gaz it will cost.

At last but not least, blockchain is a very complicated domain. Please be indulgent and considere me as a newbe in this domain. I have from far a lot of to learn!!!

references:

See:
[zero proof knoledge](https://www.youtube.com/watch?v=z7V830zndoA)

**Comment by impishlucy** [comment]
This is Godot, not Blockchain, while this is a good idea, why come with Blockchain ? 
Couldve given a Propsal about Zero Proof Knowledge without mentioning this Godforsaken Crypto Garbage.

**Reply by gogo2464** [reply]
Hi @ElryGH thank you very very much for your attention! You are a great help!

I agree with you on the point that web3, blockchain *buzzwords* are now commercial garbages as it used to be with p2p.

I suggest that I will based on the venerable freenet protocol to do my stuff.

I will improove modularity. I will create a godot c# *package*. I must NOT include it in native engine!

**Reply by gogo2464** [reply]
With blockchain I meant *cryptography*. Sorry for the confusion. I am a newbie in cryptography.

**Reply by impishlucy** [reply]
You even linked a blockchain video, you clearly meant blockchain

**Reply by niceEli** [reply]
if your making a C# package for anti cheat, why is it in proposals

**Comment by hpvb** [comment]
Using zero proof knowledge for anti-cheat in games is a really cool idea! I'm not sure how actionable that would be as a built-in Godot feature though.

As for the other people in the thread, zero proof knowledge is not a blockchain thing, necessarily. And the idea isn't ridiculous, although with my knowledge of how this works I don't see how to do it. But all that really means is that this isn't my field.

Good luck with your system though! I'll be closing this proposal as this really doesn't fit inside the engine. But if you want to maintain it as a module or gdextension it could go on the asset library!

**Comment by augus1990** [comment]
I'm here because I'm designing a P2P online game and this ZKP feature in Godot would be really useful for me. It would allow the game to check players encounters with revealing the real position of both players.
